salt/_pillar/metalk8s_nodes.py

```diff
@@ -50,13 +53,102 @@ def node_info(node, ca_minion):
     return result
 
 
+def get_cluster_version(api_client):
+    coreV1 = kubernetes.client.CoreV1Api(api_client=api_client)
+    try:
+        namespace = coreV1.read_namespace(name="kube-system")
+    except (ApiException, HTTPError) as exc:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to read namespace information {}'.format(exc)
+        )
+    annotation_key = 'metalk8s.scality.com/cluster-version'
+
+    if annotation_key not in namespace.metadata.annotations:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to retrieve cluster version'
+        )
+
+    return namespace.metadata.annotations[annotation_key]
+
+
+def iso_timestamp_converter(timestamp):
+    if not timestamp:
+        return None
+    return timestamp.isoformat()
+
+
+def get_storage_classes(api_client):
+    storageV1 = kubernetes.client.StorageV1Api(api_client=api_client)
+    storage_classes = {}
+    try:
+        storageclass_list = storageV1.list_storage_class()
+    except (ApiException, HTTPError) as exc:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to list storage classes {}'.format(exc)
+        )
+    for storageclass in storageclass_list.items:
+        # Need to convert the datetime object in storageclass to ISO format in
+        # order to make them serializable.
+        storageclass.metadata.creation_timestamp = iso_timestamp_converter(
+            storageclass.metadata.creation_timestamp
+        )
+        storageclass.metadata.deletion_timestamp = iso_timestamp_converter(
+            storageclass.metadata.deletion_timestamp
+        )
+        storage_classes[storageclass.metadata.name] = storageclass.to_dict()
+
+    return storage_classes
+
+
+def list_volumes(api_client, minion_id):
+    results = {}
+    customObjectApi = kubernetes.client.CustomObjectsApi(api_client=api_client)
+    storage_classes = get_storage_classes(api_client)
+
+    if not storage_classes:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Cannot list storage classes'
+        )
+
+    try:
+        volumes = customObjectApi.list_cluster_custom_object(
+            group="storage.metalk8s.scality.com",
+            version="v1alpha1",
+            plural="volumes"
+        )
+    except (ApiException, HTTPError) as exc:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to get volumes {}'.format(exc)
+        )
+
+    for volume in volumes['items']:
+        name = volume['metadata']['name']
+
+        if volume['spec']['nodeName'] != minion_id:
+            continue
+        storageclass = storage_classes.get(
+            volume['spec'].pop('storageClassName')
+        )
+        if not storageclass:
```

Comment:
I'd rewrite this again, to get rid of the `continue` below:

```python
if storageclass:
    volume['spec']['storageCass'] = storageClass
    result = volume
else:
    result = 'Invalid StorageClass'

results[name] = result
```

However, now there's a problem: we end up with values in the `results` dict (and as such in the pillar) that are (error) strings, and do not look like `Volume` objects. This makes writing SLSs using this pillar value *much* more cumbersome.

As such, instead of putting this error string in the pillar (especially repeating the list of StorageClasses in it, which seems quite superfluous?), I'd instead put `None` as value for the corresponding name (key in the dict), and log a proper error message.